components/eamxx/src/physics/mam/eamxx_mam_constituent_fluxes_interface.cpp

```diff
@@ -0,0 +1,292 @@
+    const char *int_nmr_field_name = mam_coupling::int_aero_nmr_field_name(m);
+    wet_aero_.int_aero_nmr[m] =
+        get_field_out(int_nmr_field_name).get_view<Real **>();
+
+    // cloudborne aerosol tracers of interest: number (n) mixing ratios
+    const char *cld_nmr_field_name = mam_coupling::cld_aero_nmr_field_name(m);
+    wet_aero_.cld_aero_nmr[m] =
+        get_field_out(cld_nmr_field_name).get_view<Real **>();
+
+    for(int a = 0; a < mam_coupling::num_aero_species(); ++a) {
+      // (interstitial) aerosol tracers of interest: mass (q) mixing ratios
+      const char *int_mmr_field_name =
+          mam_coupling::int_aero_mmr_field_name(m, a);
+
+      if(strlen(int_mmr_field_name) > 0) {
+        wet_aero_.int_aero_mmr[m][a] =
+            get_field_out(int_mmr_field_name).get_view<Real **>();
+      }
+
+      // (cloudborne) aerosol tracers of interest: mass (q) mixing ratios
+      const char *cld_mmr_field_name =
+          mam_coupling::cld_aero_mmr_field_name(m, a);
+      if(strlen(cld_mmr_field_name) > 0) {
+        wet_aero_.cld_aero_mmr[m][a] =
+            get_field_out(cld_mmr_field_name).get_view<Real **>();
+      }
+    }
+  }
+  for(int g = 0; g < mam_coupling::num_aero_gases(); ++g) {
+    const char *gas_mmr_field_name = mam_coupling::gas_mmr_field_name(g);
+    wet_aero_.gas_mmr[g] =
+        get_field_out(gas_mmr_field_name).get_view<Real **>();
+  }
+
+  //-----------------------------------------------------------------
+  // Setup preprocessing and post processing
+  //-----------------------------------------------------------------
+  preprocess_.initialize(ncol_, nlev_, wet_atm_, dry_atm_);
+
+}  // end initialize_impl()
+
+// ================================================================
+//  RUN_IMPL
+// ================================================================
+void MAMConstituentFluxes::run_impl(const double dt) {
+  const auto scan_policy = ekat::ExeSpaceUtils<
+      KT::ExeSpace>::get_thread_range_parallel_scan_team_policy(ncol_, nlev_);
+
+  // -------------------------------------------------------------------
+  // (LONG) NOTE: The following code is an adaptation of cflx.F90 code in
+  // E3SM. In EAMxx, all constituents are considered "wet" (or have wet
+  // mixing ratios), we are *not* doing any wet to dry conversions in the
+  // "preprocess" for this process. We are simply updating the MAM4xx
+  // tracers using the "constituent fluxes".
+  // We are converting wet atm to dry atm. Since we do not use or update
+  // any of the water constituents (qc, qv, qi etc.), we should be okay
+  // to do this conversion. We need to do this conversion as our function
+  // are built following HAERO data structures.
+  // -------------------------------------------------------------------
+
+  // preprocess input -- needs a scan for the calculation of dry_atm_, wet_aero_
+  // etc.
+  Kokkos::parallel_for("preprocess", scan_policy, preprocess_);
```

Comment:
Are you foreseeing preprocess to become a bigger kernel? If not, I'd consider using it here with a KOKKOS_LAMBDA. It makes the hpp much shorter, and doesn't add much more code here. Also, please use a more descrictive label, so when we use profiling tools, we can recognize it.

```c++
auto lambda = KOKKOS_LAMBDA (const Kokkos::TeamPolicy<KT::ExeSpace>::member_type &team)
      const int i = team.league_rank();  // column index
      compute_dry_mixing_ratios(team, wet_atm_pre_, dry_atm_pre_, i);
      team.team_barrier();
      // vertical heights has to be computed after computing dry mixing ratios
      // for atmosphere
      compute_vertical_layer_heights(team, dry_atm_pre_, i);
      compute_updraft_velocities(team, wet_atm_pre_, dry_atm_pre_, i);
};
Kokkos::parallel_for("mam_cfi_preprocess", scan_policy, lambda);
```

At this point, maybe using a better name than "preprocess" would also be good. "mam_cfi_updraft"?

Reply:
Removed the whole preprocess block and replaced it with the code you suggested.